src/tools/replay.cpp

```diff
@@ -83,8 +83,7 @@ std::unique_ptr<Element> pop_next_action(std::string expected_action_type) {
 	Element* clone = next_action->Clone()->ToElement();
 	root->RemoveChild(next_action);
 	
-	std::unique_ptr<Element> ptr { clone };
-	return ptr;
+	return *clone;
```

Comment:
If this is what you're doing, do you really need to clone? Couldn't you just replace the clone line by `Element clone = *next_action;`?

Reply:
Calling RemoveChild() on next_action was freeing the data. Cloning is necessary to keep it around without leaving it in the document object

Reply:
Are you saying that `Element clone = *next_action` does not copy it?

But as it is now, isn't the clone still being leaked?

Reply:
when a block does

```
{
   Element next = pop_next_action()
}
```

I thought the block is taking ownership of the data, then when the scope ends, it gets deleted.

Reply:
> Are you saying that Element clone = *next_action does not copy it?

Maybe it would.

Reply:
> I thought the block is taking ownership of the data, then when the scope ends, it gets deleted.

That's correct, but I'm talking about the clone _inside_ the function. That's a pointer, so it's not going to be automatically deleted.

Reply:
I think I get what you're saying. Let me fix that, then

Reply:
For it to not crash, Clone() is required.

`Element clone = *(next_action->Clone()->ToElement());`

This works. Does that avoid a memory leak?

Reply:
I think calling `Clone()` in and of itself _might_ be a memory leak, but I'd have to see the internals of `Clone()` be sure. The _typical_ internals of a function called clone is essentially `return new ClassName(*this)`, and if that's what it's doing, then this line is certainly a memory leak. But maybe `Clone()` is actually doing something else. I'm not sure.

Reply:
I looked into Clone() and it's definitely a memory leak. Which brings me full-circle to what I was trying to accomplish by returning a `unique_ptr`: I thought wrapping the pointer returned by Clone() in unique_ptr would make the `unique_ptr` manage that memory, so when the pointer went out of scope in the calling context of `pop_next_action()`, it would delete the clone.

Reply:
I think the document object might still manage the memory of the pointer returned by `Clone()` though. Because I get a segfault when `unique_ptr` manages the pointer and frees it.

Reply:
> I looked into Clone() and it's definitely calling `new`. Which brings me full-circle to what I was trying to accomplish by returning a `unique_ptr`: I thought wrapping the pointer returned by Clone() in unique_ptr would make the `unique_ptr` manage that memory, so when the pointer went out of scope in the calling context of `pop_next_action()`, it would delete the clone.

This is a more or less correct line of reasoning. The only problem is, it's not clear whether deleting the clone is the correct way to dispose of it.